File: Project/Simulator/Organisms/Animals/Antelope.java
package Simulator.Organisms.Animals;

import Simulator.Organisms.Organism;
import Utils.RNG;
import Simulator.World;
import Simulator.Tile;
import java.util.List;
import java.util.ArrayList;

import org.json.JSONObject;

public class Antelope extends Animal {

    public Antelope(World world) {
        super(4, 4, world, Type.ANTELOPE);
    }

    public Antelope(JSONObject json, World world) {
        super(json, world);
    }

    @Override

    public Animal construct() {
        return new Antelope(world);
    }

    @Override
    public void action() {
        super.action();
        if (tile.getOrganismCount() > 1)
            if (tile.getOrganism().isAlive())
                return;
        List<Tile> neighbours = new ArrayList<>(tile.getNeighbours());
        neighbours.remove(oldTile);
        if (neighbours.isEmpty())
            return;
        Tile newTile = neighbours.get(RNG.roll(0, neighbours.size() - 1));
        move(newTile);
    }

    @Override
    public void collision(Organism other) {
        if (other instanceof Antelope)
            super.collision(other);
        else if (RNG.roll(0, 100) < 50) {
            Tile newTile = tile.getRandomFreeNeighbour();
            if (newTile == null) {
                super.collision(other);
                return;
            }
            move(newTile);
            world.addLog(getSymbol() + " escaped from " + other.getSymbol() + "!");
        } else
            super.collision(other);
    }

    @Override
    public boolean collisionReaction(Organism other) {
        if (other instanceof Antelope)
            return false;
        if (RNG.roll(0, 100) < 50) {
            Tile newTile = tile.getRandomFreeNeighbour();
            if (newTile == null)
                return false;
            move(newTile);
            return true;
        }
        return false;
    }
}

File: Project/Simulator/Organisms/Animals/Antelope.py
from .Animal import Animal
from Simulator.Organisms.Organism import Type, Organism

import random


class Antelope(Animal):
    def __init__(self, world, json=None):
        super().__init__(4, 4, world, Type.ANTELOPE, json)

    def construct(self):
        return Antelope(self._world)

    def action(self):
        super().action()
        if self._tile.getOrganismCount() > 1:
            organism = self._tile.getOrganism()
            if organism is not None and organism.isAlive():
                return
        neighbours = list(self._tile.getNeighbours())
        if self._oldTile is None:
            return
        neighbours.remove(self._oldTile)  # type: ignore
        if not neighbours:
            return
        new_tile = neighbours[random.randint(0, len(neighbours) - 1)]
        self.move(new_tile)

    def collision(self, other: Organism):
        if other.getType() == Type.ANTELOPE:
            super().collision(other)
        elif random.randint(0, 100) < 50:
            newTile = self._tile.getRandomFreeNeighbour()
            if newTile is None:
                super().collision(other)
                return
            self.move(newTile)
            self._world.addLog(
                f"{self.getSymbol()} escaped from "f"{other.getSymbol()}!")
        else:
            super().collision(other)

    def collisionReaction(self, other):
        if other.getType() == Type.ANTELOPE:
            return False
        if random.randint(0, 100) < 50:
            newTile = self._tile.getRandomFreeNeighbour()
            if newTile is None:
                return False
            self.move(newTile)
            return True
        return False
